### PR TITLE
Feature/loss cleanup

### PR DESCRIPTION
**Pull request by d4n-hewitt**
Cleaning up the losses classes to make them more extensible, and implementing one loss, e.g. mse with test cases which can be extended

### CODE REVIEW

File: nova/src/losses/loss.py [hunk 1]
```diff
@@ -1,44 +1,57 @@
...
-            Reduced loss value.
+            Reduced loss value scalar.
         """
-        if reduction_method == "mean":
-            return (
-                values / values.size if values.size > 0 else 0.0
-            )  # TODO: implement mean in Fusion backend
-        elif reduction_method == "sum":
-            return values.sum()
+        if self.reduction_method == "mean":
```
Comment (pytergeist) [inline nova/src/losses/loss.py]: don't hardcode this here, take a look at standerdise loss in keras here: https://github.com/keras-team/keras/blob/master/keras/src/losses/loss.py

I also don't like this completely though, but I like the abstraction of a method with the responsibility to ensure the reduction method exists and throw if not. instead of hardcoding a dictionary within fn scope, lets define a dataclass that defines the allowed reduction methods -

File: nova/src/losses/loss.py [hunk 2]
```diff
@@ -1,44 +1,57 @@
...
+            loss = values.sum() / values.size
+        elif self.reduction_method == "mean_with_sample_weight":
+            if sample_weights is None:
```
Comment (pytergeist) [inline nova/src/losses/loss.py]: raising errors deserves it's own method with a single responsibility to raise errors

File: nova/src/losses/loss.py [hunk 3]
```diff
@@ -1,44 +1,57 @@
...
+                raise ValueError("sample weights empty")
+            loss = values.sum() / sample_weights.sum()
```
Comment (pytergeist) [inline nova/src/losses/loss.py]: this should be more generically define once the dataclass with allowed reduction methods is set up

**Review by pytergeist — COMMENTED** [review]
few comments

**Review by pytergeist — APPROVED** [review]
lgtm
